// secretjs/dist/grpc_gateway/confio/proofs.pb.js
"use strict";
/* eslint-disable */
// @ts-nocheck
/*
* This file is a generated Typescript file for GRPC Gateway, DO NOT MODIFY
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LengthOp = exports.HashOp = void 0;
var HashOp;
(function (HashOp) {
    HashOp["NO_HASH"] = "NO_HASH";
    HashOp["SHA256"] = "SHA256";
    HashOp["SHA512"] = "SHA512";
    HashOp["KECCAK"] = "KECCAK";
    HashOp["RIPEMD160"] = "RIPEMD160";
    HashOp["BITCOIN"] = "BITCOIN";
})(HashOp = exports.HashOp || (exports.HashOp = {}));
var LengthOp;
(function (LengthOp) {
    LengthOp["NO_PREFIX"] = "NO_PREFIX";
    LengthOp["VAR_PROTO"] = "VAR_PROTO";
    LengthOp["VAR_RLP"] = "VAR_RLP";
    LengthOp["FIXED32_BIG"] = "FIXED32_BIG";
    LengthOp["FIXED32_LITTLE"] = "FIXED32_LITTLE";
    LengthOp["FIXED64_BIG"] = "FIXED64_BIG";
    LengthOp["FIXED64_LITTLE"] = "FIXED64_LITTLE";
    LengthOp["REQUIRE_32_BYTES"] = "REQUIRE_32_BYTES";
    LengthOp["REQUIRE_64_BYTES"] = "REQUIRE_64_BYTES";
})(LengthOp = exports.LengthOp || (exports.LengthOp = {}));
