// secretjs/dist/grpc_gateway/ibc/lightclients/solomachine/v1/solomachine.pb.js
"use strict";
/* eslint-disable */
// @ts-nocheck
/*
* This file is a generated Typescript file for GRPC Gateway, DO NOT MODIFY
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataType = void 0;
var DataType;
(function (DataType) {
    DataType["DATA_TYPE_UNINITIALIZED_UNSPECIFIED"] = "DATA_TYPE_UNINITIALIZED_UNSPECIFIED";
    DataType["DATA_TYPE_CLIENT_STATE"] = "DATA_TYPE_CLIENT_STATE";
    DataType["DATA_TYPE_CONSENSUS_STATE"] = "DATA_TYPE_CONSENSUS_STATE";
    DataType["DATA_TYPE_CONNECTION_STATE"] = "DATA_TYPE_CONNECTION_STATE";
    DataType["DATA_TYPE_CHANNEL_STATE"] = "DATA_TYPE_CHANNEL_STATE";
    DataType["DATA_TYPE_PACKET_COMMITMENT"] = "DATA_TYPE_PACKET_COMMITMENT";
    DataType["DATA_TYPE_PACKET_ACKNOWLEDGEMENT"] = "DATA_TYPE_PACKET_ACKNOWLEDGEMENT";
    DataType["DATA_TYPE_PACKET_RECEIPT_ABSENCE"] = "DATA_TYPE_PACKET_RECEIPT_ABSENCE";
    DataType["DATA_TYPE_NEXT_SEQUENCE_RECV"] = "DATA_TYPE_NEXT_SEQUENCE_RECV";
    DataType["DATA_TYPE_HEADER"] = "DATA_TYPE_HEADER";
})(DataType = exports.DataType || (exports.DataType = {}));
