// secretjs/dist/extensions/snip1155/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSnip1155ChangeMetadata = exports.MsgSnip1155RemoveMinter = exports.MsgSnipAddMinter = exports.MsgSnip1155BatchTransfer = exports.MsgSnip1155Transfer = exports.MsgSnip1155Burn = exports.MsgSnip1155Mint = exports.MsgSnip1155BatchSend = exports.MsgSnip1155Send = exports.MsgSnip1155RemoveCurator = exports.MsgSnip1155AddCurator = exports.MsgSnip1155CurateTokens = exports.MsgSnip1155RemoveAdmin = exports.MsgSnip1155ChangeAdmin = void 0;
const tx_1 = require("../../tx");
class MsgSnip1155ChangeAdmin extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155ChangeAdmin = MsgSnip1155ChangeAdmin;
class MsgSnip1155RemoveAdmin extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155RemoveAdmin = MsgSnip1155RemoveAdmin;
class MsgSnip1155CurateTokens extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155CurateTokens = MsgSnip1155CurateTokens;
class MsgSnip1155AddCurator extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155AddCurator = MsgSnip1155AddCurator;
class MsgSnip1155RemoveCurator extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155RemoveCurator = MsgSnip1155RemoveCurator;
class MsgSnip1155Send extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155Send = MsgSnip1155Send;
class MsgSnip1155BatchSend extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155BatchSend = MsgSnip1155BatchSend;
class MsgSnip1155Mint extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155Mint = MsgSnip1155Mint;
class MsgSnip1155Burn extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155Burn = MsgSnip1155Burn;
class MsgSnip1155Transfer extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155Transfer = MsgSnip1155Transfer;
class MsgSnip1155BatchTransfer extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155BatchTransfer = MsgSnip1155BatchTransfer;
class MsgSnipAddMinter extends tx_1.MsgExecuteContract {
}
exports.MsgSnipAddMinter = MsgSnipAddMinter;
class MsgSnip1155RemoveMinter extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155RemoveMinter = MsgSnip1155RemoveMinter;
class MsgSnip1155ChangeMetadata extends tx_1.MsgExecuteContract {
}
exports.MsgSnip1155ChangeMetadata = MsgSnip1155ChangeMetadata;
