// secretjs/dist/grpc_gateway/secret/compute/v1beta1/types.pb.js
"use strict";
/* eslint-disable */
// @ts-nocheck
/*
* This file is a generated Typescript file for GRPC Gateway, DO NOT MODIFY
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractCodeHistoryOperationType = exports.AccessType = void 0;
var AccessType;
(function (AccessType) {
    AccessType["UNDEFINED"] = "UNDEFINED";
    AccessType["NOBODY"] = "NOBODY";
    AccessType["ONLY_ADDRESS"] = "ONLY_ADDRESS";
    AccessType["EVERYBODY"] = "EVERYBODY";
})(AccessType = exports.AccessType || (exports.AccessType = {}));
var ContractCodeHistoryOperationType;
(function (ContractCodeHistoryOperationType) {
    ContractCodeHistoryOperationType["CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED"] = "CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED";
    ContractCodeHistoryOperationType["CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT"] = "CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT";
    ContractCodeHistoryOperationType["CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE"] = "CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE";
    ContractCodeHistoryOperationType["CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS"] = "CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS";
})(ContractCodeHistoryOperationType = exports.ContractCodeHistoryOperationType || (exports.ContractCodeHistoryOperationType = {}));
